From 03c3f44e1db5bd37c30806f43e528a6d17df6b2c Mon Sep 17 00:00:00 2001
From: Michael Tupek <mrtupek@gmail.com>
Date: Sun, 17 May 2026 16:44:01 -0600
Subject: [PATCH 3/6] Fix SparseMatrix leak in DerivativeAssemble

Only allocate a new SparseMatrix when A is nullptr. Previously, every
Assemble() call unconditionally did A = new SparseMatrix(...), leaking
the prior allocation in Newton iteration loops.

Fixes bug 3.
---
 fem/dfem/backends/local_qf/derivative_assemble.hpp | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/fem/dfem/backends/local_qf/derivative_assemble.hpp b/fem/dfem/backends/local_qf/derivative_assemble.hpp
index 39f2852e9a..d1cba01f0a 100644
--- a/fem/dfem/backends/local_qf/derivative_assemble.hpp
+++ b/fem/dfem/backends/local_qf/derivative_assemble.hpp
@@ -296,7 +296,10 @@ struct DerivativeAssemble
       }, num_entities, thread_blocks, shmem_info.total_size, shmem_cache.ReadWrite());
 
       // Create sparse matrix and add element matrices
-      A = new SparseMatrix((*test_fes)->GetVSize(), (*trial_fes)->GetVSize());
+      if (A == nullptr)
+      {
+         A = new SparseMatrix((*test_fes)->GetVSize(), (*trial_fes)->GetVSize());
+      }
 
       auto Ae_host = Reshape(Ae_mem.HostReadWrite(),
                              num_test_dof * test_vdim,
-- 
2.39.5 (Apple Git-154)

